x/mongo/driver/batch_cursor.go

```diff
@@ -308,25 +308,39 @@ func (bc *BatchCursor) KillCursor(ctx context.Context) error {
 	}.Execute(ctx)
 }
 
-func (bc *BatchCursor) getMore(ctx context.Context) {
-	bc.clearBatch()
-	if bc.id == 0 {
-		return
-	}
+// getMoreBatchSize returns the number of documents that should be returned in
+// the response of a "getMore" operation given the limit, batchSize, and number
+// of documents already returned.
+func (bc *BatchCursor) getMoreBatchSize(ctx context.Context) (int32, error) {
+	gmBatchSize := bc.batchSize
 
-	// Required for legacy operations which don't support limit.
-	numToReturn := bc.batchSize
+	// Account for legacy operations that don't support setting a limit.
 	if bc.limit != 0 && bc.numReturned+bc.batchSize >= bc.limit {
-		numToReturn = bc.limit - bc.numReturned
-		if numToReturn <= 0 {
+		gmBatchSize = bc.limit - bc.numReturned
+		if gmBatchSize <= 0 {
 			err := bc.Close(ctx)
 			if err != nil {
-				bc.err = err
+				return 0, err
 			}
-			return
 		}
 	}
 
+	return gmBatchSize, nil
+}
+
+func (bc *BatchCursor) getMore(ctx context.Context) {
+	bc.clearBatch()
+	if bc.id == 0 {
+		return
+	}
+
+	numToReturn, err := bc.getMoreBatchSize(ctx)
```

Comment:
This refactor changes the behavior of `getMore` if the "limit" emulation overrides the batch size to be <= 0. Previously, `getMore` would close the cursor and return. Now, it closes the cursor, sets `numToReturn` to 0, and continues to run the operation.

Reply:
Really good catch, thank you. I've updated the refactored function to return an "ok" value. The reason for breaking out the function was principally for unit testing.